banditpylib/protocols/collaborative_protocol.py

```diff
@@ -0,0 +1,138 @@
+from typing import List, cast, Dict, Tuple
+from copy import deepcopy as dcopy
+
+import numpy as np
+from absl import logging
+
+from banditpylib.bandits import Bandit
+from banditpylib.data_pb2 import Trial, Actions
+from banditpylib.learners import Learner
+from banditpylib.learners.collaborative_learner import CollaborativeBAILearner
+from .utils import Protocol
+
+
+class CollaborativeLearningProtocol(Protocol):
+  """Collaborative learning protocol :cite:`tao2019collaborative`
+
+  This protocol is used to simulate the multi-agent game
+  as discussed in the paper. It runs in rounds. During each round,
+  the protocol runs the following steps in sequence:
+
+  * fetch the state of the environment and ask each learner for actions;
+  * send the actions to the enviroment for execution;
+  * update each learner with the corresponding feedback of the environment;
+  * repeat the above steps until every agent enters the WAIT or STOP state;
+  * if there is at least one agent in WAIT state, then receive information
+    broadcasted from every waiting agent, and use it to decide
+    the parameters of the next round.
```

Comment:
... then receive information broadcasted from every waiting agent and send them to master to decide arm assignment of next round.